llvm/lib/YkIR/YkIRWriter.cpp

```diff
@@ -699,6 +699,19 @@ class YkIRWriter {
         if (Attr.getKindAsEnum() == Attribute::AllocSize) {
           continue;
         }
+
+        if (Attr.getKindAsEnum() == Attribute::Alignment) {
+          // Following what we do with loads/stores, we accept any alignment
```

Comment:
This is problematic: we don't deal correctly with alignment bigger than the size of the object; we only deal correctly with equal to the size of the object AFAIK. So I think the line below (and loads/stores) should be `=` rather than `>=`.

Reply:
My understanding is that we musn't straddle an alignment boundary, which can't happen as long as the size of the object isn't bigger than the alignment. This came from a discussion with LLVM devs a few months back and ultimately lead to [this comment explaining in the serialisation of loads/stores](https://github.com/ykjit/ykllvm/blob/fa46a94c9c346b684301a2764e6c26e89a11acf1/llvm/lib/YkIR/YkIRWriter.cpp#L876-L885).

[This section of the LLVM language reference](https://llvm.org/docs/LangRef.html#id211) talks about it too.

If the `>=` is incorrect (and you explain it to me) then we can tighten what we let through.

Reply:
OK. On Monday I'll change the serialiser to reject all non-natural alignments.

Reply:
Thanks! It would be good to catch these now, but I wonder if we've already been letting some through?!

Reply:
OK, so [44fcf1a](https://github.com/ykjit/ykllvm/pull/213/commits/44fcf1a30ed75ad1cfb4a0b46943337ecdc6d0f2) rejects non-natural alignment on function parameters.

[cf01cbf](https://github.com/ykjit/ykllvm/pull/213/commits/cf01cbf0a1a8c4d2c04a5aada2ed025da7e63c76) does the same for load and store instructions.

The latter commit breaks yklua because it does things like:
```
%3184 = load i8, ptr %3183, align 8
```

I've reviewed and thought about what alignment really means for the JIT.

There are two kinds of memory location we have to consider:
 - Memory allocated, and therefore aligned by, the AOT compiler.
 - Memory allocated by JITted code.

In the world of LLVM, most memory allocations are performed by either an `alloca` instruction, or by a call to `malloc`.

 - The trace builder currently doesn't have a handler for the `alloca` instruction, which means the JIT is never allocating stack memory with special alignment requirements.
 - `malloc'd memory is always naturally aligned.

(there are some more obscure allocation functions that can align stuff in a non-natural way like `alloc_aligned` and `mmap` that we should probably crash if we encounter)

From this I deduce that when we encounter an instruction like `%3184 = load i8, ptr %3183, align 8`:

 - The AOT codegen has ensured it's aligned `8`
 - It fits in a register, so it can be done in one write (e.g. x64 `mov`) without fear of straddling an alignment boundary, since the alignment is greater than the size of the object. For the above load example, how could you place a single byte on an 8-byte boundary that also straddles the next alignment boundary? You can't.

So with the addition of some assertions I think what we were doing before is in fact (not ideal, but) correct.

If we don't buy this reasoning though, then it's time to go back and encode alignment into our AOT IR and pass it on to the JIT IR, so the JIT codegen can reason out this stuff in an explicit fashion.

Thoughts?

Reply:
> load i8, ptr %3183, align 8

Why is this rejected? It's 8 bytes (a pointer) and aligned to 8 bytes isn't it?

Reply:
The instruction says "load a byte (i8) from this pointer, and it's address is aligned to 8 bytes".

Reply:
> it can be done in one read without fear of straddling an alignment boundary, since the alignment is greater than the size of the object

I've just found a [sentence in the LLVM language reference](https://llvm.org/docs/LangRef.html#id211) that explains what I'm trying to convey here:

>  An alignment value higher than the size of the loaded type implies memory up to the alignment value bytes can be safely loaded without trapping in the default address space.

Reply:
> There are two kinds of memory location we have to consider:
> *     Memory allocated, and therefore aligned by, the AOT compiler.
> *    Memory allocated by JITted code.

Right, good point. A question is: how do we know we're doing the right thing with memory allocated by the JIT -- which includes things spilled to the stack? I'm not on top form today, so I'm struggling to think through the consequences here.

Reply:
> A question is: how do we know we're doing the right thing with memory allocated by the JIT -- which includes things spilled to the stack?

It's a good question.

We've already established that for now the JIT isn't allocating any stack space with `alloca` instructions, so we can forget that (for now).

For malloc'd memory is allocated by an external facility, so I think we can ignore this as long as we abide by the load/store align requirements upon those pointers.

The spill case is one I hadn't considered and I had to think hard.

 - A llvm local variable can't have its address taken after the fact. If you need the address of something local to a frame, you'd have to `alloca` it and you'd get a pointer in return. The `alloca` of, `load`s from and `store`s to this pointer would communicate alignment requirements explicitly at the IR level, but crucially the register allocator isn't "managing" the memory behind the alloca.

 -  A non-alloca'd local variable (a.k.a. an SSA variable) and has no requirement to live in memory at all. If a local variable spills in JITted code, then the only consumer of the spilled memory should be the unspill code of the JIT's register allocator itself. The allocator would be the only party blessed with the knowledge that a local variable spilled. So the spill/unspill code in the JIT's RA is free to mandate its own alignment requirements separate to those of `alloca` memory in the LLVM IR and it (as far as I can see) has no need to communicate what they were to the wider world.

Do you buy this reasoning?

Reply:
That does sound plausible, so let's go with it. What's the right thing to do with this PR now?

Reply:
I *think* we've convinced ourselves that non-natural alignments where the alignment requirement is greater than or equal to the size of the object are OK. If that's the case the last two commits can be reverted. Is that OK?

(Reflection: For me this conversation has really highlighted that explicit alignment in the IR would be much easier to reason about. If for nothing else, because the locality of reasoning can be fully-contained in the JIT codegen -- right now you have to think about what the serialiser allows and how those properties flow through the IRs to the codegen)